components/channel/src/channel.c

```diff
@@ -0,0 +1,143 @@
+#include "channel.h"
+#include "channel_internal.h"
+
+#include <string.h>
+#include <freertos/projdefs.h>
+
+/**
+ * CHANNEL_INIT - Macro to provide initialization for channel object
+ */
+#define CHANNEL_INIT(name, identifier, ctx, flags, callback) { "" identifier "", LIST_HEAD_INIT(name.same), LIST_HEAD_INIT(name.unique), ctx, flags, callback }
+
+/**
+ * CHANNEL - Macro to create and initialize channel object
+ */
+#define CHANNEL(name, identifier, ctx, flags, callback) Channel name = CHANNEL_INIT(name, "" identifier "", ctx, flags, callback)
+
+static CHANNEL(root_channel, "", NULL, 0, NULL);
+
+void
+channel_init
+(Channel *ch, const char *identifier, void *ctx, const BaseType_t flags, Channel_callback callback)
+{
+    channel_internal_init(ch, identifier, ctx, flags, callback);
+    channel_internal_register(ch);
+}
+
+EXPORT
+void
+channel_internal_init 
+(Channel *ch, const char *identifier, void *ctx, const BaseType_t flags, Channel_callback callback)
+{
+    ch->identifier = identifier;
+    INIT_LIST_HEAD(&ch->unique);
+    INIT_LIST_HEAD(&ch->same);
+    ch->ctx = ctx;
+    ch->flags = flags;
+    ch->callback = callback;
+}
+
+void
+channel_reset
+(Channel *ch)
+{
+    channel_internal_unregister(ch);
+}
+
+EXPORT
+void 
+channel_internal_register
+(Channel *ch) 
+{
+    Channel *curr;
+    list_for_each_entry(curr, &root_channel.unique, unique) {
+        if (strcmp(ch->identifier, curr->identifier) == 0) {
+            list_add(&ch->same, &curr->same);
+            return;
+        }
+    }
+
+    list_add(&ch->unique, &root_channel.unique);
+}
+
+EXPORT
+void 
+channel_internal_unregister
+(Channel *ch) 
+{
+    int isUnique = ! list_empty(&ch->unique);
+    int isSame   = ! list_empty(&ch->same);
```

Comment:
why are some variables camel case now?

```suggestion
    int is_unique = ! list_empty(&ch->unique);
    int is_same   = ! list_empty(&ch->same);
```